### DISCUSSION TITLE
Berry serial problem

### DISCUSSION THREAD

**Discussion by storchalberto** [arendst/Tasmota#21111] [General]
The .close() in the serial class do not function: "the 'serial' object has no attribute 'close'

**Comment by s-hadinger** [comment]
What version are you using? This is a recent addition

**Reply by storchalberto** [reply]
14.4.0 (ESP32) customized using benzino77

**Reply by sfromis** [reply]
You need latest development, as `serial.close()` was added less than two weeks ago. https://github.com/arendst/Tasmota/pull/21042

**Reply by storchalberto** [reply]
ok, you know if the update is loaded inside the benzino77 enviroment?

**Reply by sfromis** [reply]
Not using the tool, but https://github.com/benzino77/tasmocompiler says that you can select `development`

**Reply by storchalberto** [reply]
he says, but It is not true. The documentation says that is possible to select the developement, but in the dropdown it is not present....I thik that if you produce a new release, it will be integrated in benzino

**Reply by barbudor** [reply]
![IMG_20240406_181521](https://github.com/arendst/Tasmota/assets/2996042/9ddf1df5-767f-468c-aebc-161d61ebbc6d)

If you have more questions about TasmoCompiler, ask there

**Reply by sfromis** [reply]
Did you try clicking the link suggested?

![image](https://github.com/arendst/Tasmota/assets/47082390/3ffaab17-63b0-4f8e-8ea3-5d1ac70842a6)

**Reply by storchalberto** [reply]
sorry but i do not understand. I have used thr link you suggested: https://github.com/benzino77/tasmocompiler, but i do not see the image. I see the standard page of benzino 77 that allow to customize the version, but the dropdown final that allow to select the version, do not allow to select development.

**Reply by sfromis** [reply]
The image I posted is near the bottom of the page. Did you try clicking the link given there?

**Reply by storchalberto** [reply]
OK thanks, the link function well and the close() now function

**Reply by storchalberto** [reply]
I use in the program some constant values. I have defined this in a var (example: var myconst=123). exist a statement that allow to define constant values?

**Reply by sfromis** [reply]
If what you mean is not constant, but persistent across boots, you have `Mem` variables in addition to `Var`. Another approach is boot rules setting such values.

**Reply by storchalberto** [reply]
Sorry, probably im am not clear. What i means is the equivalent of the const statement of javascript:
const PI = 12;
if (myval==PI)
{
var myvar=PI+1
}

**Reply by sfromis** [reply]
Maybe you mean https://tasmota.github.io/docs/Berry/#persist-module

**Reply by storchalberto** [reply]
No, i am talking about constant values written inside the source code. The same as the example i show in javascript, or as EQU assigment in assembler language, or const in php language.

**Reply by sfromis** [reply]
https://github.com/berry-lang/berry/wiki/Chapter-6#class-variables-static

**Reply by storchalberto** [reply]
I see but static is usable only inside a class. i need a definition global in the whole program

**Reply by sfromis** [reply]
No problem. Just refer to it via the class, from the "outside".

But you are really not at all clear about what your concerns are, compared to just allocating a normal global variable. If you do not choose to overwrite the content, it is a constant.

**Reply by storchalberto** [reply]
OK i see the solution:
class globals
	static p1=1
	static p2=2
end
print(globals.p1)

**Reply by storchalberto** [reply]
Ok, i do not know the way berry interpreter is written. I suppose that if i define a var, it needs a memory allocation in ram, non dependant if i modify or not the value. If i define a constant value, it do not require ram memory. I am wrong?

**Reply by sfromis** [reply]
Yes, you are wrong. **Anything** you do needs memory allocation, including but not limited to a var. Code also needs to be in RAM. No way around that. Just like in JavaScript. OTOH, in some languages/compilers, constructs like `const` may help the compiler to optimize code.

What I meant with a simple global variable was not to create a class named "globals", but use the existing feature of the Berry language. If you outside of a class/instance scope create a variable, this will implicitly be global, also accessible from classes. You can also do it explicitly, like `global.p2=2` (as `global` is already defined). Global variables created implicitly will also be accessible as members of `global`, and vice versa. Anytime you create a class or function (from the global level), this is also "just" becoming a global variable housing the code.

The benefit of doing it in a class is to avoid "polluting" the global name space, which can complicate things when the amount of code grows. That makes it more clean to have definitions in a class, but you are not bound by such principles of style. Thus, I'm not saying anything against your "globals" class, just trying to clarify what I meant.

What you don't get in Berry is to be blocked from changing the value. If you really want this, you could create a class instance with virtual members, where you then could cause the code to fail upon attempts to modify a value. That would probably be over-complicating things for now....

**Reply by storchalberto** [reply]
Thanks. now it is clear for me.
